File: python/src/day11/main.py
from src.common.file_utils import get_path, read_lines


def part_one(filename: str) -> int:
    lines = read_lines(get_path(__file__, filename))
    neighbors = [(1, 0), (1, -1), (0, -1), (-1, -1),
                 (-1, 0), (-1, 1), (0, 1), (1, 1)]
    area = [[c for c in line] for line in lines]
    h, w = len(area), len(area[0])
    
    while True:
        new_area = [[area[r][c] for c in range(w)] for r in range(h)]
        for r in range(h):
            for c in range(w):
                occupied = 0

                for n in neighbors:
                    nr, nc = r + n[0], c + n[1]
                    if 0 <= nr < h and 0 <= nc < w and area[nr][nc] == "#":
                        occupied += 1

                if area[r][c] == "L" and occupied == 0:
                    new_area[r][c] = "#"

                elif area[r][c] == "#" and occupied >= 4:
                    new_area[r][c] = "L"

        if new_area == area:
            return sum([x.count("#") for x in area])
        else:
            area = new_area

    return -1


def part_two(filename: str) -> int:
    lines = read_lines(get_path(__file__, filename))
    neighbors = [(1, 0), (1, -1), (0, -1), (-1, -1),
                 (-1, 0), (-1, 1), (0, 1), (1, 1)]
    area = [[c for c in line] for line in lines]
    h, w = len(area), len(area[0])
    
    while True:
        new_area = [[area[r][c] for c in range(w)] for r in range(h)]
        for r in range(h):
            for c in range(w):
                occupied = 0

                for n in neighbors:
                    nr, nc = r + n[0], c + n[1]
                    while 0 <= nr < h and 0 <= nc < w:
                        if area[nr][nc] == "#":
                            occupied += 1
                            break
                        elif area[nr][nc] == "L":
                            break
                        nr, nc = nr + n[0], nc + n[1]

                if area[r][c] == "L" and occupied == 0:
                    new_area[r][c] = "#"

                elif area[r][c] == "#" and occupied >= 5:
                    new_area[r][c] = "L"

        if new_area == area:
            return sum([x.count("#") for x in area])
        else:
            area = new_area

    return -1


if __name__ == '__main__':

    print("---Part One---")
    part_one_res = part_one("input.txt")
    print(part_one_res)

    print("---Part Two---")
    part_two_res = part_two("input.txt")
    print(part_two_res)

File: go/day11/main.go
package main

import (
	input "aoc2020/inpututils"

	"fmt"
	"strconv"
	"strings"
)

const (
	emptySeat    rune = 'L'
	occupiedSeat rune = '#'
	path         rune = '.'
)

var (
	directions = []pos{pos{-1, -1}, pos{-1, 0}, pos{-1, 1}, pos{0, -1}, pos{0, 1}, pos{1, -1}, pos{1, 0}, pos{1, 1}}
)

type pos struct {
	r, c int
}

type tiles map[pos]rune

type floor struct {
	tiles         tiles
	height, width int
}

func main() {
	fmt.Println("--- Part One ---")
	fmt.Println(Part1("test1.txt"))

	fmt.Println("--- Part Two ---")
	fmt.Println(Part2("input.txt"))
}

func Part1(filename string) int {
	floorString := input.ReadRaw(filename)
	f := newFloor(floorString)
	changes := true
	for changes == true {
		f, changes = f.getNext()
	}
	return f.countOccupied()
}

func Part2(filename string) int {
	floorString := input.ReadRaw(filename)
	f := newFloor(floorString)
	changes := true
	for changes == true {
		f, changes = f.getNext2()
	}
	return f.countOccupied()
}

func newFloor(floorString string) floor {
	lines := strings.Split(floorString, "\n")

	t := tiles{}
	h := len(lines)
	w := len(lines[0])

	for r, s := range lines {
		for c, tile := range s {
			p := pos{r, c}
			t[p] = tile
		}
	}
	return floor{
		height: h,
		width:  w,
		tiles:  t,
	}
}

func (f floor) getNext() (floor, bool) {
	changes := false
	newFloor := floor{
		height: f.height,
		width:  f.width,
		tiles:  tiles{},
	}
	for r := 0; r < f.height; r++ {
		for c := 0; c < f.width; c++ {
			loc := pos{r, c}
			tile := f.tiles[loc]
			if tile == path {
				newFloor.tiles[loc] = tile
				continue
			}
			occupied := f.countOccupiedNeighbours(loc)
			switch {
			case tile == emptySeat && occupied == 0:
				newFloor.tiles[loc] = occupiedSeat
				changes = true
			case tile == occupiedSeat && occupied >= 4:
				newFloor.tiles[loc] = emptySeat
				changes = true
			default:
				newFloor.tiles[loc] = tile
			}
		}
	}
	return newFloor, changes
}

func (f floor) getNext2() (floor, bool) {
	changes := false
	newFloor := floor{
		height: f.height,
		width:  f.width,
		tiles:  tiles{},
	}
	for r := 0; r < f.height; r++ {
		for c := 0; c < f.width; c++ {
			loc := pos{r, c}
			tile := f.tiles[loc]
			if tile == path {
				newFloor.tiles[loc] = tile
				continue
			}
			occupied := f.countOccupiedNeighboursInSight(loc)
			switch {
			case tile == emptySeat && occupied == 0:
				newFloor.tiles[loc] = occupiedSeat
				changes = true
			case tile == occupiedSeat && occupied >= 5:
				newFloor.tiles[loc] = emptySeat
				changes = true
			default:
				newFloor.tiles[loc] = tile
			}
		}
	}
	return newFloor, changes
}

func (f floor) draw() string {
	s := ""
	for r := 0; r < f.height; r++ {
		for c := 0; c < f.width; c++ {
			loc := pos{r, c}
			s += string(f.tiles[loc])
		}
		s += "\n"
	}
	return s
}

func (f floor) countOccupied() int {
	occupied := 0
	for r := 0; r < f.height; r++ {
		for c := 0; c < f.width; c++ {
			if f.tiles[pos{r, c}] == occupiedSeat {
				occupied += 1
			}
		}
	}
	return occupied
}

func (f floor) countOccupiedNeighbours(l pos) int {
	var occupied int
	for _, dir := range directions {
		r, c := l.r+dir.r, l.c+dir.c
		if (0 <= c && c < f.width) && (0 <= r && r < f.height) {
			if f.tiles[pos{r, c}] == occupiedSeat {
				occupied += 1
			}
		}
	}
	return occupied
}

func (f floor) countOccupiedNeighboursInSight(l pos) int {
	var occupied int
	for _, dir := range directions {
		r, c := l.r+dir.r, l.c+dir.c
		for (0 <= c && c < f.width) && (0 <= r && r < f.height) && f.tiles[pos{r, c}] != emptySeat {
			if f.tiles[pos{r, c}] == occupiedSeat {
				occupied += 1
				break
			}
			r, c = r+dir.r, c+dir.c
		}
	}
	return occupied
}

func check(err error) {
	if err != nil {
		panic(err)
	}
}

func toInt(s string) int {
	i, err := strconv.Atoi(s)
	check(err)
	return i
}
